content/runner.js

```diff
@@ -0,0 +1,106 @@
+// Copyright 2015, 2016 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+/* global chrome */
+import uuid from 'uuid/v4';
+
+let instance = null;
+
+export default class Runner {
+
+  message = {};
+  port = null;
+
+  constructor () {
+    this.setup();
+    instance = this;
+  }
+
+  static get () {
+    if (!instance) {
+      return new Runner();
+    }
+
+    return instance;
+  }
+
+  setup () {
+    // Setup a Promise-based communication with the background process
+    this.port = chrome.runtime.connect({ name: 'id' });
+    this.messages = {};
+
+    // Listen for responses
+    this.port.onMessage.addListener((msg) => this.handleIncomingMessage(msg));
+  }
+
+  static execute (task, data) {
+    return Runner.get().execute(task, data);
+  }
+
+  execute (task, input) {
+    const id = uuid();
```

Comment:
Each `port` connection is separate in background thread, so there is no issue with collisions here.

Why do we send only one request? I thought we will send a request per match? If we wait for everything befoere doing anythnig it might take super long if there is a lot of addresses, I think it would be better to do it async for each element.